src/libraries/Common/src/System/Net/Http/aspnetcore/Quic/Implementations/Managed/Internal/QuicSocketContext.cs

```diff
@@ -0,0 +1,218 @@
+                }
+            }
+
+            await Task.Yield();
+
+            byte[] recvBuffer = new byte[64 * 1024]; // max UDP packet size
+            var reader = new QuicReader(recvBuffer);
+            byte[] sendBuffer = new byte[64 * 1024]; // max UDP packet size
+            var writer = new QuicWriter(sendBuffer);
+
+            var socket = new Socket(_localEndpoint.AddressFamily, SocketType.Dgram, ProtocolType.Udp);
+            if (_localEndpoint != null)
+            {
+                socket.Bind(_localEndpoint);
+            }
+
+            Task<SocketReceiveFromResult> socketReceiveTask =
+                socket.ReceiveFromAsync(recvBuffer, SocketFlags.None, _localEndpoint);
+
+            Task[] waitingTasks = {socketReceiveTask, _signalTcs.Task};
+
+            try
+            {
+                while (!token.IsCancellationRequested)
+                {
+                    await Task.WhenAny(waitingTasks);
+
+                    if (socketReceiveTask.IsCompleted)
+                    {
+                        var result = await socketReceiveTask;
+
+                        // discard too small datagrams as they cannot be valid packets
+                        if (result.ReceivedBytes >= QuicConstants.MinimumPacketSize)
+                        {
+                            reader.Reset(recvBuffer, 0, result.ReceivedBytes);
+                            var connection = DispatchPacket(reader, (IPEndPoint) result.RemoteEndPoint);
+                            if (connection != null)
+                            {
+                                // also query the connection for data to be sent back
+                                await SendData(connection!, writer, socket, sendBuffer);
+                            }
+                        }
+
+                        // start new receiving task
+                        waitingTasks[0] = socketReceiveTask =
+                            socket.ReceiveFromAsync(recvBuffer, SocketFlags.None, _localEndpoint);
+                    }
+
+                    if (_signalTcs.Task.IsCompleted)
+                    {
+                        _signalTcs = new TaskCompletionSource<int>();
+                        waitingTasks[1] = _signalTcs.Task;
+
+                        if (_client != null)
+                        {
+                            await SendData(_client, writer, socket, sendBuffer);
+                        }
+                        else
+                        {
+                            foreach (ManagedQuicConnection connection in _connections!.Values)
+                            {
+                                await SendData(connection, writer, socket, sendBuffer);
+                            }
+                        }
+                    }
+
+                }
+            }
+            catch (Exception e)
+            {
+                Console.WriteLine(e);
```

Comment:
If you wish to log then I suggest: `NetEventSource`, example for http: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/NetEventSource.Http.cs and then look for usages, e.g.: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Net.Http/src/System/Net/Http/HttpClient.cs#L658

Reply:
Thanks for the examples, another item on my list.